### PR TITLE
Oxidize core event enum

### LINKED ISSUE #3796: Oxidize `CoreEvent` enum

### PR DESCRIPTION
**Pull request by Volham22**
# What has changed ?

This close #3796 

- [ ] Does this PR affect the User ?

### CODE REVIEW

File: src/enumerate.rs [hunk 1]
```diff
@@ -642,3 +642,436 @@ impl UserProfile {
...
 }
+
+#[pyclass]
+pub(crate) struct CoreEvent(libparsec::core::CoreEvent);
+
+#[pymethods]
+impl CoreEvent {
+    #[classattr]
+    #[pyo3(name = "BACKEND_CONNECTION_CHANGED")]
+    fn backend_connection_changed() -> &'static PyObject {
```
Comment (TimeEngineer) [inline src/enumerate.rs:653]: Maybe we can have a macro now to generate these block

File: oxidation/libparsec/crates/core/src/events.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+#[derive(Clone, PartialEq, Eq, Debug)]
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core/src/events.rs:3]: Can you put `Debug` first ? I like to have essential derive come first

File: src/enumerate.rs (hunk 1 above)
Comment (Volham22) [reply src/enumerate.rs:653]: Sure, I've used vim macros myself to generate this code.

Comment (Volham22) [reply src/enumerate.rs]: Done.

File: oxidation/libparsec/crates/core/src/events.rs (hunk 2 above)
Comment (Volham22) [reply oxidation/libparsec/crates/core/src/events.rs]: Done.

File: src/enumerate.rs (hunk 1 above)
Comment (TimeEngineer) [reply src/enumerate.rs]: I thought about a generic macro which can be used by others :thinking:

Comment (Volham22) [reply src/enumerate.rs]: I modified the macro to make it more generic. I think we should use this enum in an other PR.

File: src/binding_utils.rs [hunk 3]
```diff
@@ -176,8 +176,30 @@ macro_rules! create_exception {
     };
 }
 
+macro_rules! impl_enum_field {
+    ($pyo3_name: literal, $fn_name: ident, $enum_class: ident, $field_value: expr) => {
```
Comment (TimeEngineer) [inline src/binding_utils.rs:180]: `enum_class` should come first, because you implement for.
And why don't you group in one macro ?

Comment (Volham22) [reply src/binding_utils.rs]: I forgot that patterns in rust macros can be repeated. I modified the macro, now a single invocation can generate all fields.
```rs
crate::binding_utils::impl_enum_field!(
    CoreEvent,
    [
        "BACKEND_CONNECTION_CHANGED",
        backend_connection_changed,
        libparsec::core::CoreEvent::BackendConnectionChanged
    ],
   // ...
);
```

**Review by TimeEngineer — APPROVED** [review]

**Review by TimeEngineer — APPROVED** [review]

**Pull request merged**